### DISCUSSION TITLE
Decimate does not change matrix size (returns bool false)

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1818] [Q&A]
I have STL files with fine mesh which I want to decimate for the subsequent mesh boolean operation (because of the computation time needed). I tried using `igl::decimate() & igl::qslim()`, but both return `bool false` value. I tried using different maximum face counts and on different parts. But neither of them are working and as a result, the vertex/face matrix sizes don't change and mesh remains as is. 

I used the command line operation `MeshStatistics` to check if something is wrong with the mesh (edge orientation, non-manifold edges etc.), but nothing seems wrong. Also, when I debugged into the `igl::decimate()` function, I found that it breaks from the `while loop` with the condition `e == -1` and hence the bool variable `clean_finish = false` (line 218 of `decimate.cpp`). 

What could be the reason for this ? For coarse meshes, `igl::upsample()` works correctly on the parts that I have, but `igl::decimate()` doesn't on fine meshes. Do I need to provide some custom `cost_and_placement & stopping_condition` function callbacks ?

**Comment by alecjacobson** [comment]
Are you snapping the vertices together (e.g., with `igl::remove_duplicate_vertices` before calling `igl::decimate`? .stl files do not store meshes, rather lists of disconnected triangles.

If your problems persist, please upload a problematic .stl and the snippet of code reproducing your problem.

**Reply by drathin4** [reply]
I tried both, with/without `igl::remove_duplicate_vertices()` before `igl::decimate()`, but that doesn't make any difference. The final result is the same - it returns `bool false`.

Following is the code snippet. Both `igl::decimate() & igl::qslim()` don't seem to work.

![image](https://user-images.githubusercontent.com/79855313/120074338-fc576200-c0b9-11eb-838b-a2f0ad7d983f.png)

I'm attaching the problematic *.stl file.

[02_BoxModule.zip](https://github.com/libigl/libigl/files/6564837/02_BoxModule.zip)

**Reply by alecjacobson** [reply]
```
mergestl /Users/ajx/Downloads/02_BoxModule.{stl,ply}
meshstatistics /Users/ajx/Downloads/02_BoxModule.ply
```

reveals

```
filename                     /Users/ajx/Downloads/02_BoxModule.ply
number of edges                                             373050
number of faces                                             248700
number of vertices                                          124352
number of dimensions                                             3
bounding box diagonal                                      24.0821
minimum angle                                          1.83952e-05
maximum angle                                              3.13376
minimum area                                                 2e-09
maximum area                                               19.1358
volume                                                     -299.03
centroid_x                                                -131.212
centroid_y                                                -45.7678
centroid_z                                                 91.7117
number of small triangles                                      184
number of small angles                                         960
number of close vertices                                         0
number of connected components                                   1
number of unreferenced vertices                                  0
number of handles                                                0
Euler characteristic                                             2
number of boundary loops                                         0
number of boundary edges                                         0
number of nonmanifold edges                                      0
number of conflictedly oriented edges                          1242
number of duplicate vertices                                     0
number of combinatorially duplicate faces                        0
number of geometrically degenerate faces                         0
number of combinatorially degenerate faces                       0
number of intra-component self-intersecting pairs                3
number of self-intersecting pairs                                3
```

In particular,
```
number of conflictedly oriented edges                         1242
```
Same type problem as your [previous issue](https://github.com/libigl/libigl/discussions/1806#discussioncomment-750378)

**Reply by drathin4** [reply]
OK, I saw that you did `mergestl` before using `meshstatistics`. I had checked the `meshstatistics` without using `mergestl`, hence I didn't find any conflicting edges. However I had done `igl::remove_duplicate_vertices()` before calling `igl::decimate()`. 

So I think, using `meshfix` seems the only solution to this problem.

**Comment by alecjacobson** [comment]
All mergestl does is call remove_duplicate_vertices
https://github.com/alecjacobson/gp-cli/blob/master/mergestl.cpp

In any case, yes, you might need to call meshfix. Ideally, you could reach
upstream and figure out why you're getting these conflicting edges in the
first place. If you have access to how these meshes are made.

>

**Reply by drathin4** [reply]
I have a couple of questions on this topic:

1. I used the `MeshFix` command through command line on each of the files that I have. But I noticed that it increases the file size by 3x - 5x. Why does that happen ? Does it make the mesh finer ?
2. I tried using `igl::decimate()` on a mesh-fixed .stl file. But still I'm having the same problem that I had when I created this issue (if you check at the top). It doesn't alter the matrix size and return value is `bool false`. What could be the reason, considering that I have already ascertained that there are no conflictedly oriented edges ? I will attach a sample .stl file for you to check that it has been mesh-fixed. I used the commands in the following order to reduce the file size after mesh-fixing: 
MeshFix - mergestl - meshstatistics (to confirm) - convertmesh. Ideally I'm expecting the `V, F` matrix sizes to reduce and `AverageEdgeLength` to increase after using `igl::decimate()`.

**Comment by alecjacobson** [comment]
Please post your meshfix questions to meshfix GitHub repo.

If you run your mesh through meshstatistics do you see problematic edges?
If not then post the mesh.

**Reply by drathin4** [reply]
I can't find any problematic edges after running `MeshStatistics`. In the previous reply, I forgot to send the .stl file, so attaching it now.

[02_BoxModule_fixed_remdupvtcs.zip](https://github.com/libigl/libigl/files/6570714/02_BoxModule_fixed_remdupvtcs.zip)

**Reply by alecjacobson** [reply]
Thanks for sharing the file. This is a .stl file, so you _must_ merge the vertices before treating it as a connected mesh. Decimate appears to work fine on this mesh:

```
mergestl /Users/ajx/Downloads/02_BoxModule_fixed_remdupvtcs.{stl,ply}
decimate 1000 /Users/ajx/Downloads/02_BoxModule_fixed_remdupvtcs{,-1000}.ply
scrubmesh /Users/ajx/Downloads/02_BoxModule_fixed_remdupvtcs{,-1000}.ply
```

![scrubmesh-000000](https://user-images.githubusercontent.com/2241689/120211412-eacba280-c1fe-11eb-8be3-059a1555dda5.png)
![scrubmesh-000001](https://user-images.githubusercontent.com/2241689/120211452-f1f2b080-c1fe-11eb-9695-1f2457ded0cd.png)

In libigl this corresponds to calling igl::remove_duplicate_vertices _before_ igl::decimate / igl::qslim.

**Reply by jdumas** [reply]
Also try to avoid saving meshes as .stl. Prefer to use .obj or .ply, which both preserve connectivity information.

**Reply by drathin4** [reply]
I was using `igl::decimate()` directly on STL files and hence it was not working. I didn't realize, we have to merge the vertices before trying to decimate the mesh. My input is all STL files, hence I was making all operations directly on STL files. Henceforth I'll try using PLY format. Thanks for the information.
